### DISCUSSION TITLE
Gatsby v4: `jsonId` automatic property prevents Json mapping in `gatsby-config` to be done

### DISCUSSION THREAD

**Discussion by no-chris** [gatsbyjs/gatsby#33798] [Help] [answered]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

It seems that Gatsby v4 breaks the `json` mapping because of the new `jsonId` property mentioned here: https://www.gatsbyjs.com/docs/reference/release-notes/migrating-from-v3-to-v4/#gatsby-transformer-json

As I understand, that is because the mapping between the 2 entities is done by default on the `id` property. Since this property is now replaced by `jsonId`, the mapping is not done anymore.

Replacing `id` with `jsonId` here seems to do the trick, although the fix is probably more subtle: https://github.com/gatsbyjs/gatsby/blob/1a73b017e7d5d267986ca07bdf7f7cd454797aa2/packages/gatsby/src/schema/infer/add-inferred-fields.js#L217

### Reproduction Link

https://github.com/no-chris/json-mapping-bug

### Steps to Reproduce

1. Declare a mapping between 2 json files exactly as described here https://www.gatsbyjs.com/docs/reference/config-files/gatsby-config/#mapping-node-types
2. Run the following query
```graphql
query MyQuery {
  experience: allExperienceJson {
    edges {
      node {
        company
        position
        from
        to
        items {
          label
          description
          tech {
            icon
            color
            label
          }
        }
      }
    }
  }
}
```

### Expected Result

```json
{
  "data": {
    "experience": {
      "edges": [
        {
          "node": {
            "company": "Company A",
            "position": "Unicorn Developer",
            "from": "Dec 2016",
            "to": "Present",
            "items": [
              {
                "label": "Responsibility",
                "description": "Being an unicorn",
                "tech": null
              },
              {
                "label": "Hands on",
                "description": null,
                "tech": [
                  {
                    "icon": "facebook",
                    "color": "teal",
                    "label": "React"
                  },
                  {
                    "icon": "server",
                    "color": "green",
                    "label": "NodeJS"
                  }
                ]
              }
            ]
          }
        }
      ]
    }
  },
  "extensions": {}
}
```

### Actual Result

Notice `"tech": []`
```json
{
  "data": {
    "experience": {
      "edges": [
        {
          "node": {
            "company": "Company A",
            "position": "Unicorn Developer",
            "from": "Dec 2016",
            "to": "Present",
            "items": [
              {
                "label": "Responsibility",
                "description": "Being an unicorn",
                "tech": null
              },
              {
                "label": "Hands on",
                "description": null,
                "tech": []
              }
            ]
          }
        }
      ]
    }
  },
  "extensions": {}
}
```

### Environment

```shell
System:
    OS: macOS 11.4
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.18.0 - ~/.nvm/versions/node/v14.18.0/bin/node
    npm: 6.14.15 - ~/.nvm/versions/node/v14.18.0/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 95.0.4638.54
    Firefox: 91.0.2
    Safari: 14.1.1
  npmPackages:
    gatsby: ^4.0.2 => 4.0.2
    gatsby-source-filesystem: ^4.0.0 => 4.0.0
    gatsby-transformer-json: ^4.0.0 => 4.0.0
  npmGlobalPackages:
    gatsby-cli: 4.0.0
```

**Comment by LekoArts** [comment, marked as answer]
Hi!

You'll need to update your usage of `mapping` to make this work:

![image](https://user-images.githubusercontent.com/16143594/139698896-8d43b7a4-7dba-464a-8fc3-34320c0e7bf7.png)

```js
mapping: {
  'ExperienceJson.items.tech': `TechJson.jsonId`
}
```

I'll update the doc to update that and highlight again that people should use `@link` directive via schema customization instead: https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#foreign-key-fields

So for your case this would be:

```js
exports.createSchemaCustomization = ({ actions }) => {
  const { createTypes } = actions

  createTypes(`
    type ExperienceJsonItems {
      tech: [TechJson] @link(by: "jsonId")
    }
  `)
}
```

You can use `__typename` to get the names you need:

![image](https://user-images.githubusercontent.com/16143594/139701743-23fefa37-b1f7-4079-9398-56ae8c549dc5.png)

---

PR: https://github.com/gatsbyjs/gatsby/pull/33799

**Comment by no-chris** [comment]
That worked like a charm for my use case, thanks a lot! 🙌🏻
